### ISSUE TITLE
Views fail to index with timeout prior to os_process_timeout elapsing

### ISSUE THREAD

**Issue by javamonn** [apache/couchdb#1298]
I'm running into an issue getting an instance of CouchDB stood back up after a restart. It was running fine prior to the restart, but now seems to be running into an issue when building the views from scratch. I've since played with the configuration a bit to try to debug.

You can see the full logs of my situation [here](https://gist.github.com/javamonn/ed8ee8c778471c18ba1fb13433ed3300), with some sensitive bits redacted out.

Some notable lines from the logs:
- os_process_timeout is set to 600000, an absurdly high, not real-world value: https://gist.github.com/javamonn/ed8ee8c778471c18ba1fb13433ed3300#file-gistfile1-txt-L38
- There's an early `[os_mon]` error, prior to indexing kicking off that seems important but I can't find much information about it: https://gist.github.com/javamonn/ed8ee8c778471c18ba1fb13433ed3300#file-gistfile1-txt-L244
- Indexing seems to time out after 5 seconds, ignoring the `os_process_timeout` I've set. [begin](https://gist.github.com/javamonn/ed8ee8c778471c18ba1fb13433ed3300#file-gistfile1-txt-L244), [end](https://gist.github.com/javamonn/ed8ee8c778471c18ba1fb13433ed3300#file-gistfile1-txt-L314).

## Expected Behavior
View indexing should respect `os_process_timeout` and eventually complete, or give a more appropriate error message if there is another error.

## Current Behavior
View indexing times out prior to `os_process_timeout`, indexing in Futon diplays stuck at 0%.

## Possible Solution
Is there possibly another error occurring, and CouchDB is swallowing it with "OS process timed out."? https://issues.apache.org/jira/browse/COUCHDB-2431 seems to imply this could be the case. I've tried adding some logging to `handleError` in `main.js` to no avail, is there somewhere else where added logging or a log file would increase visibility into the error? I have SSH access to the machine and seem to be able to execute the same `couchjs` invocations that `couchdb` does.

## Context
This instance is deployed to Google Container Engine using an image close to the [official image](https://github.com/apache/couchdb-docker/blob/1a7c4254c158a194ff195da6ebfed910d24a95b7/2.1.1/Dockerfile). I can't reproduce this locally with an image close to what I have deployed, but I haven't tried locally with a backup of the production data. The instance has 5 databases (including `_users` and `_replicator`, the largest one of which is about 59MB total, with 151993 documents.

## Your Environment

* Version used: 2.1.1
* Browser Name and version: N/A
* Operating System and version (desktop or mobile): `Linux <REDACTED> 4.4.86+ #1 SMP Thu Dec 7 20:11:11 PST 2017 x86_64 GNU/Linux`
* Link to your project: N/A

**Comment by javamonn** [comment]
I've just noticed the `[cluster]` section from the logs. I'm actually only running a single node, so the default values of `n = 3` is incorrect. I've updated to `n = 1` and redeployed but that doesn't seem to have affected the outcome.

I've added some logging to main `readline()` loop in `main.js` and I'm not seeing any output on view creation, which suggests that something more fundamental is messed up with my install, as the query server is just hanging.

I've found http://docs.couchdb.org/en/latest/install/troubleshooting.html and am going through it now.

**Comment by wohali** [comment]
@javamonn 

You're right in that the cpu_sup process failing suggests that your instance is heavily CPU constrained. The 5 second timeout may mean that couchjs is failing to even start up correctly - you're not even getting an initial "hello" handshake back from that spawned process.

Are you able to increase the CPU allocation for your instance? If so, please do. CouchDB runs best with 4 or more cores allocated.

We are tracking numerous situations in which CouchDB deployed in containers is simply unreliable. We don't have enough people to debug these scenarios right now. So far, in most cases it's simply too few CPU or RAM resources being allocated.

If you need a guaranteed stable CouchDB we recommend deploying it in the traditional fashion on dedicated VMs or bare metal if possible.

**Comment by javamonn** [comment]
@wohali incredible diagnosis, bumping the CPU allocation I was giving the container did the trick. Now I'm not sure how the instance was running at all prior to the restart taking it offline.

I was giving it about a core originally. I bumped it up to 2 cores, and the `cpu_sup` error went away, but indices would still timeout after a couple of seconds. I bumped it up to 4 cores and it's humming away now.

Are the hardware recommendations documented anywhere? I found similar recommendations on [Couchbase](https://developer.couchbase.com/documentation/server/current/install/pre-install.html), but can't find any on CouchDB proper.

Thanks for the help.
